kernel/src/main/java/org/kframework/compile/RemoveUnit.java

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Runtime Verification, Inc. All Rights Reserved.
+package org.kframework.compile;
+
+import org.kframework.Collections;
+import org.kframework.attributes.Att;
+import org.kframework.definition.Module;
+import org.kframework.definition.Rule;
+import org.kframework.definition.Sentence;
+import org.kframework.kore.*;
+import org.kframework.utils.errorsystem.KEMException;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static org.kframework.definition.Constructors.*;
+import static org.kframework.kore.KORE.*;
+import static org.kframework.Collections.*;
+
+public class RemoveUnit {
+
+  private Module m;
+
+  public Module apply(Module module) {
+    m = module;
+    return Module(module.name(), module.imports(), stream(module.localSentences()).flatMap(this::gen).collect(Collections.toSet()), module.att());
+  }
+
+  private Stream<Sentence> gen(Sentence s) {
+    if (!(s instanceof Rule)) {
+      return Stream.of(s);
+    }
+    Rule r = (Rule)s;
+    K body = flattenLists(r.body());
+    K requires = flattenLists(r.requires());
+    K ensures = flattenLists(r.ensures());
+    return Stream.of(Rule(body, requires, ensures, r.att()));
+  }
+
+  private K flattenLists(K k) {
+    return new TransformK() {
+      @Override
+      public K apply(KApply k) {
+        Att att = m.attributesFor().getOrElse(k.klabel(), () -> Att.empty());
+
+        // Ignore optional cells, which have a unit attribute but no assoc
+        if (   att.contains(Att.CELL())
+            && att.contains(Att.MULTIPLICITY())
+            && att.get(Att.MULTIPLICITY()).equals("?")) {
+          return super.apply(k);
+        }
```

Comment:
Why is this check necessary?
Can't we just remove throwing the exception below?
This is an optimization step and is not critical, so I don't think we should enforce `unit` with `assoc` present.
If `unit` must have `assoc`, then maybe it should be a verification step earlier in the kompile pipeline.

Reply:
We don't have a way to flatten unit from lists that doesn't depend on the production being associative. That's why the check exists. I tend to agree that it's better to special case this one exception.

Reply:
Then the exception below should contain location information to the production.
So change `attributesFor` to `productionsFor`.
Then add a test in `regression-new/checks`.